sympy/codegen/ast.py

```diff
@@ -161,7 +161,17 @@ def _mk_Tuple(args):
 
 
 class CodegenAST(Basic):
-    pass
+    __slots__ = ()
+
+
+def _fields(cls):
+    if hasattr(cls, "_fields"):
+        yield from cls._fields
```

Comment:
Why do we need this?

Reply:
Hmmm, I did indeed forget to explain.

The reason this is needed (and we cannot only use `__slots__`) is that there is no way to 'automatically' order inherited slots in the correct way.

One sensible way seemed from in reverse MRO:

```python
class A:
    __slots__ = ('x', 'y')

class B:
    __slots__ = ('z', )

_fields(B)  # x, y, z
```

This seems to be the rule followed by most `codegen.ast` members

However, some classes (like `FunctionPrototype`) add inherited slots in a different way:

```python
class Node:
    __slots__ = ("attrs", )

class FunctionPrototype(Node):
    __slots__ = ('return_type', 'name', 'parameters') + Node.__slots__

_fields(FunctionPrototype)  # attrs, return_type, name, parameters
# SHOULD BE return_type, name, parameters, attrs
```

I briefly considered just adopting the new order, but that would of course be a breaking change.

Reply:
If you prefer, I could also make `_fields` the only way to define fields (i.e. never fallback to `__slots__`). That'd simplify introspection but it'd require setting `_fields` on a lot of classes.

Reply:
If going by the mro is causing ordering issues then I think it is indeed better to simply define `_fields` fully. It would also make it clearer what to do for a future author who wants to create another codegen ast class.